src/runtime_src/core/edge/user/shim.cpp

```diff
@@ -508,7 +509,38 @@ xclLoadAxlf(const axlf *buffer)
   drm_zocl_axlf axlf_obj = {
     .za_xclbin_ptr = const_cast<axlf *>(buffer),
     .za_flags = flags,
+    .za_ksize = 0,
+    .za_kernels = NULL,
   };
+
+  auto kernels = xrt_core::xclbin::get_kernels(buffer);
+  /* Calculate size of kernels */
+  for (auto kernel : kernels) {
+      axlf_obj.za_ksize += sizeof(kernel_info) + sizeof(argument_info) * kernel.args.size();
+  }
+
+  /* Check PCIe's shim.cpp for details of kernels binary */
+  axlf_obj.za_kernels = (char *)malloc(axlf_obj.za_ksize);
```

Comment:
Please no malloc/free/new/delete, this is not exception safe
Use std::vector<char*> data(size) to manage the buffer and extract the buffer where necessary using vec.data().

But I am not seeing a free of this buffer.  Wouldn't it be better if the driver allocated the memory and did the copying?   A dangling malloc is not acceptable.  You already have all the data allocated on the user side, so the copying done here could as well be done in the driver.  You can define a struct like kernel_info that just contains the user pointer addresses already allocated. e.g. kernel.name.c_str() which is valid for the duration of the xclLoad call so no additional strcpy is necessary.

Reply:
yes, you are right. Driver would allocate a buffer and copy this binary.
Use c++ container is much more better. :)